spec/datadog/tracing/integration_spec.rb

```diff
@@ -284,6 +291,227 @@ def agent_receives_span_step3(previous_success)
+        expect(spans).to_not include(have_tag('_dd.span_sampling.max_per_second'))
+      end
+    end
+
+    shared_examples 'set single span sampling tags' do
+      let(:rule_rate) { 1.0 }
+      let(:max_per_second) { -1 }
+
+      it do
+        expect(single_sampled_span.get_metric('_dd.span_sampling.mechanism')).to eq(8)
+        expect(single_sampled_span.get_metric('_dd.span_sampling.rule_rate')).to eq(rule_rate)
+        expect(single_sampled_span.get_metric('_dd.span_sampling.max_per_second')).to eq(max_per_second)
+      end
+    end
+
+    shared_examples 'flushed complete trace' do |expected_span_count: 3|
+      it_behaves_like 'flushed trace'
+
+      it 'flushed all spans' do
+        expect(spans).to have(expected_span_count).items
+      end
+    end
+
+    context 'with default settings' do
+      it_behaves_like 'flushed complete trace'
+      it_behaves_like 'does not modify spans'
+    end
+
+    context 'with a kept trace' do
+      let(:trace_sampling_rate) { 1.0 }
+
+      it_behaves_like 'flushed complete trace'
+      it_behaves_like 'does not modify spans'
+    end
+
+    context 'with a dropped trace' do
+      context 'by priority sampling' do
+        let(:trace_sampling_rate) { 0.0 }
+
+        context 'with rule matching' do
+          context 'on name' do
+            context 'with a dropped span' do
+              let(:rules) { [{ name: 'my.op', sample_rate: 0.0 }] }
+
+              it_behaves_like 'flushed complete trace'
+              it_behaves_like 'does not modify spans'
+
+              context 'by rate limiting' do
+                let(:rules) { [{ name: 'my.op', sample_rate: 1.0, max_per_second: 0 }] }
+
+                it_behaves_like 'flushed complete trace'
+                it_behaves_like 'does not modify spans'
+              end
+            end
+
+            context 'with a kept span' do
+              let(:rules) { [{ name: 'my.op', sample_rate: 1.0 }] }
+
+              it_behaves_like 'flushed complete trace'
+              it_behaves_like 'set single span sampling tags'
+            end
+          end
+
+          context 'on service' do
+            context 'with a dropped span' do
+              let(:rules) { [{ service: 'my-ser*', sample_rate: 0.0 }] }
+
+              it_behaves_like 'flushed complete trace'
+              it_behaves_like 'does not modify spans'
+
+              context 'by rate limiting' do
+                let(:rules) { [{ service: 'my-ser*', sample_rate: 1.0, max_per_second: 0 }] }
+
+                it_behaves_like 'flushed complete trace'
+                it_behaves_like 'does not modify spans'
+              end
+            end
+
+            context 'with a kept span' do
+              let(:rules) { [{ service: 'my-ser*', sample_rate: 1.0 }] }
+
+              it_behaves_like 'flushed complete trace'
+              it_behaves_like 'set single span sampling tags'
+            end
+          end
+        end
+      end
```

Comment:
This being an integration test, I'm not sure we need to exhaustively test all features of the rule matcher -- after all, we have tests for that, and we're just duplicating them.

Instead, I would suggest simplifying this to only testing a situation where the matches matches, and another where it doesn't. All other variants are effectively tests for the local logic in the sampler that don't otherwise reflect differently on tracing-as-a-whole.

Reply:
Hmmm, I see duplication for sure.
I tried to match the somewhat higher level requirements from the RFC here in the integrations suite.

I'll see if I can simply these without removing too much in that I'm not as confident in the feature anymore.

Reply:
Ok, I removed half the tests inside `context 'with a dropped trace'`, specifically the distinct cases where rules could match based on span name or service name: now the tests only test if the span match, the field that was matched on is not important.